web/components/roomConfig/GitlabRepoConfig.tsx

```diff
@@ -0,0 +1,167 @@
+import { h, FunctionComponent, createRef } from "preact";
+import { useState, useCallback, useEffect } from "preact/hooks";
+import BridgeAPI from "../../BridgeAPI";
+import { ConnectionConfigurationProps, RoomConfig } from "./RoomConfig";
+import { GitLabRepoConnectionState, GitLabRepoResponseItem, GitLabTargetFilter, GitLabRepoConnectionTarget, GitLabRepoConnectionProjectTarget, GitLabRepoConnectionInstanceTarget } from "../../../src/Connections/GitlabRepo";
+import InputField from "../InputField";
+import ButtonSet from "../ButtonSet";
+import { Button } from "../Button";
+
+const EventType = "uk.half-shot.matrix-hookshot.gitlab.repository";
+
+const ConnectionSearch: FunctionComponent<{api: BridgeAPI, onPicked: (state: GitLabRepoConnectionState) => void}> = ({api, onPicked}) => {
+    const [filter, setFilter] = useState<GitLabTargetFilter>({});
+    const [results, setResults] = useState<GitLabRepoConnectionProjectTarget[]|null>(null);
+    const [instances, setInstances] = useState<GitLabRepoConnectionInstanceTarget[]|null>(null);
+    const [debounceTimer, setDebounceTimer] = useState<number>(null);
+    const [currentProjectPath, setCurrentProjectPath] = useState<string|null>(null);
+
+    const searchFn = useCallback(() => {
+        api.getConnectionTargets<GitLabRepoConnectionTarget>(EventType, filter).then(res => {
+            if (!filter.instance) {
+                setInstances(res);
+                if (res[0]) {
+                    setFilter({instance: res[0].name});
+                }
+            } else {
+                setResults(res as GitLabRepoConnectionProjectTarget[]);
+            }
+        });
+    }, [api, filter]);
+
+    const updateSearchFn = useCallback((evt: InputEvent) => {
+        setFilter({...filter, search: (evt.target as HTMLInputElement).value });
+    }, [filter, setFilter]);
+
+    useEffect(() => {
+        if (debounceTimer) {
+            clearTimeout(debounceTimer);
+        }
+        // Browser types
+        setDebounceTimer(setTimeout(searchFn, 500) as unknown as number);
+        // Things break if we depend on the thing we are clearing.
+        // eslint-disable-next-line react-hooks/exhaustive-deps
+    }, [setDebounceTimer, searchFn, filter, instances]);
+
+    useEffect(() => {
+        const hasResult = results?.find(n => n.name === filter.search);
+        if (hasResult) {
+            onPicked(hasResult.state);
+            setCurrentProjectPath(hasResult.state.path);
+        }
+    }, [onPicked, results, filter, setCurrentProjectPath]);
+
+
+
+    return <div>
+        {instances === null && <p> Loading GitLab instances. </p>}
+        {instances?.length === 0 && <p> You are not logged into any GitLab instances. </p>}
+        <InputField visible={instances?.length > 0} label="GitLab Instance" noPadding={true}>
```

Comment:
Why even render this with `visible = false` if there are no instances?

Reply:
Not sure I follow? This doesn't render if there are no instances. The visible thing is a shorthand to avoid doing `{ doWeWantToShowThis && <input` everywhere, since it's a common need for input fields.

Reply:
Ah, I see.. You've intentionally added a `visible` prop to some components to make the syntax of the parent component cleaner.

I like `{foo &&` better, but I see that it's mainly personal preference of what's considered cleaner.

Reply:
Yeah, I suppose it is more uniform. Maybe I'm a bit of a holdout from the Angular era :), either way, let's leave it as it is for the moment.